iwyu.cc

```diff
@@ -2118,43 +2102,54 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
         if (!current_ast_node()->template HasAncestorOfType<CallExpr>() ||
             ContainsKey(
                 GetCallerResponsibleTypesForAutocast(current_ast_node()),
-                deref_to_type)) {
-          need_full_deref_to_type = true;
+                RemovePointersAndReferences(to_type))) {
+          required_full_types.push_back(to_type);
         }
         break;
       // Need the full from-type so we can call its 'operator <totype>()'.
       case clang::CK_UserDefinedConversion:
-        need_full_deref_from_type = true;
+        required_full_types.push_back(from_type);
         break;
 
       // Kinds that cast up or down an inheritance hierarchy.
       case clang::CK_BaseToDerived:
       case clang::CK_BaseToDerivedMemberPointer:
         // Just 'to' type is enough: full type for derived gets base type too.
-        need_full_deref_to_type = true;
+        required_full_types.push_back(to_type);
         break;
       case clang::CK_DerivedToBase:
       case clang::CK_UncheckedDerivedToBase:
       case clang::CK_DerivedToBaseMemberPointer:
-        need_full_deref_from_type = true;
+        // Just 'from' type is enough: full type for derived gets base type too.
+        required_full_types.push_back(from_type);
+
+        // If this derived-to-base cast requires a conversion function, it's a
+        // user-defined conversion operator. The from-type in this cast is not
+        // necessarily related to the base type, but the converted-to type
+        // must be.
+        if (converted_to_type) {
+          required_full_types.push_back(converted_to_type);
+        }
```

Comment:
I think clang will unwrap the casts one be one and reach the `CK_UserDefinedConversion` eventually. This should be moved there to cover all cases, not just those inside a `DerivedToBase`

Reply:
There was some complication about that... Yes --

This cast: `static_cast<Derived*>(Convertible())` will only require the full type for `Convertible`, not the converted-to-type, `Derived*` -- a fwd decl use is enough there. This is the expression we see in `CK_UserDefinedConversion`.

However, this cast: `static_cast<Base*>(&derived)` will require the full type for both `Base` and `Derived` because the cast needs to know how they're related.

So the combination of the two -- `static_cast<Base*>(Convertible())` will only see a cast from `Convertible` to `Derived*` for `CK_UserDefinedConversion`, and a cast from `Derived*` to `Base*` for `CK_DerivedToBase`.

There's no single cast expression that sees the context of all these, that's why the conversion function is a nice indicator that the derived-to-base cast needs more than the standard stuff.

Reply:
Oh got it. I had the feeling I might be missing something :)

I intuitively thought that we would always want a full-use of the *to* type for a user-defined conversion, but yeah, I guess it depends a lot on what you actually do with it.

Maybe you could mention why we need the *to* type here? Might also help to explain above that the *converted_to_type* might be useful for other cast kinds in the future, so that it warrants a separate definition.

Reply:
The function doc comment goes into some detail, but I'll expand a little on these `Type`s too, because it took me far too long to get my head around it.

Reply:
Wait, you ask why we need the "to type". We don't -- we need the *from* type, both the actual from type, and the from type after any user-defined conversions.

I wonder if the name `converted_**to**_type` is misleading -- I meant for that to read "the type we user-defined-converted to", but it sounds like you're interpreting it as "converted to-type"?

I think the comments explain what's going on pretty explicitly (improvement suggestions welcome here, if you have them!), but I'll find a better name for `converted_to_type` to try and reduce the confusion.

Reply:
That's how I read it, yes. I think it is clearer now. Thanks!

Reply:
:+1: